src/pgrn-normalize.c

```diff
@@ -0,0 +1,58 @@
+#include "pgroonga.h"
+#include "pgrn-compatible.h"
+#include "pgrn-global.h"
+#include "pgrn-groonga.h"
+
+#include <utils/builtins.h>
+
+static grn_ctx *ctx = &PGrnContext;
+
+PGRN_FUNCTION_INFO_V1(pgroonga_normalize);
+
+/**
+ * pgroonga_normalize(target text, normalizerName text) : text
+ */
+Datum
+pgroonga_normalize(PG_FUNCTION_ARGS)
+{
+	text *target;
+	text *normalizerName;
+	grn_obj *normalizer;
+	grn_obj *string;
+	unsigned int length_in_bytes;
+	const char *normalized;
+	text *ret;
+
+	target = PG_GETARG_TEXT_PP(0);
+
+	if (PG_NARGS() == 2) {
+		normalizerName = PG_GETARG_TEXT_PP(1);
+		normalizer = PGrnLookupWithSize(VARDATA_ANY(normalizerName),
+										VARSIZE_ANY_EXHDR(normalizerName),
+										ERROR);
+	} else {
+		normalizer = PGrnLookup(PGRN_DEFAULT_NORMALIZER, ERROR);
+	}
+
+	string = grn_string_open(ctx,
+							 VARDATA_ANY(target),
+							 VARSIZE_ANY_EXHDR(target),
+							 normalizer,
+							 0);
+	if (!string)
```

Comment:
Probably you should return immediately in this case.

Reason:

* NULL pointer will be stored in `normalized` in this case.
    * https://github.com/groonga/groonga/blob/master/lib/string.c#L248
* `normalized` (which is a `NULL` pointer) will be passed to `cstring_to_text_with_len()`
* cstring_to_text_with_len() will pass it to `memcpy()`
    * https://doxygen.postgresql.org/builtins_8h.html#a95beff77197e26615b6e8f120ba285a3
* Passing `NULL` pointer to `memcpy()` may causes undefined behavior:
    * https://stackoverflow.com/questions/5243012/is-it-guaranteed-to-be-safe-to-perform-memcpy0-0-0

Reply:
Because `ereport` will abort the execution immediately in this case, we don't need to place a `return` statement here. For details, please refere to [the official document](https://www.postgresql.org/docs/9.6/static/error-message-reporting.html):

> If the severity level is ERROR or higher, ereport aborts the execution of the user-defined function and does not return to the caller. If the severity level is lower than ERROR, ereport returns normally.

Reply:
Sorry, it's wrong.

`ereport` will call `longjmp` when the error code is `ERROR`:

* https://doxygen.postgresql.org/elog_8h.html

Your code is no problem.